gatereader/reader.py

```diff
@@ -0,0 +1,187 @@
+                for j in range(0, 2):
+                    self._transmit_packet(PacketId.RX_ERROR, bytes())
+                    try:
+                        p, payload = self._expect_packet(resp_length)
+                        if p.id == PacketId.RX_ERROR:
+                            # RxError was a response to our RxError. This
+                            # behaviour is unacceptible, let's punish the
+                            # reader by resetting it!
+                            raise Reader.ReaderError()
+                        else:
+                            return (p, payload)
+                    except (Reader.CorruptedPacketException,
+                            serial.SerialTimeoutException):
+                        # Again, corrupted response. Retry
+                        pass
+        # If we got here, we've had 3 consectutive failed receive
+        # attempts. Reader, how dare you! We will reset him for doing
+        # this to us!
+        raise Reader.ReaderError()
+
+    def _check_atr(self):
+        if self.port.inWaiting() == 5:
+            # Trouble is waiting for us in the buffer; there should be
+            # nothing there
+            # HACK ALERT: We must release sooner than we are ready.
+            # This listens for ACK instead of the ATR as the preliminary
+            # version does not have a bootloader
+            try:
+                self._expect_packet(5)
+            except serial.SerialTimeoutException:
+                # Now we are in serious trouble!
+                raise ReaderError()
+            raise ResetException()
+
+    def set_leds(self, leds):
+        self._check_atr()
+        head, payload = self._retransmit_wrapper(
+            PacketId.SET_LED,
+            bytes([leds]),
+            ResponseLength.RESPONSE_ATR.value
+        )
+        if head.id != PacketId.ACK.value:
+            # The Reader did something it shouldn't have done; reset him
+            raise Reader.ReaderError()
+
+    def beep(self, tones, repeat=False):
+        if len(tones) > Reader.MAX_BEEP_LENGTH:
+            raise ValueError("Cannot beep more than 8 times in one command")
+        self._check_atr()
+        payload = bytearray(0)
+        for tone in tones:
+            payload += struct.pack(STRUCT_FORMAT + 'H', tone[0])
+            payload += struct.pack(STRUCT_FORMAT + 'H', tone[1])
+        if repeat:
+            payload += bytes([0x01])
+        else:
+            payload += bytes([0x00])
+        head, payload = self._retransmit_wrapper(
+            PacketId.BEEP,
+            payload,
+            ResponseLength.RESPONSE_ATR.value
+        )
+        if head.id != PacketId.ACK.value:
+            # The Reader did something it shouldn't have done; reset him
+            raise Reader.CorruptedPacketException()
+
+    def RFID_send(self, payload):
+        if len(payload) > Reader.MAX_RFID_PAYLOAD:
+            raise ValueError("Cannot transmit more than 128 bytes at once")
```

Comment:
`"Cannot transmit more than {} bytes at once".format(Reader.MAX_RFID_PAYLOAD)` to avoid duplicating the constant.